### ISSUE TITLE
`LatLonGrid` bilinear regridding returns `NaN` for queries exactly on the upper grid boundary

### ISSUE THREAD

**Issue by nbren12** [NVlabs/earth2grid#66]
`earth2grid` bilinear lat/lon regridding appears to treat the upper coordinate boundary as out-of-bounds.

A query exactly equal to the maximum source latitude produces `NaN`, even when it lies exactly on a source grid row. The lower boundary is accepted, so the behavior is asymmetric.

This seems to come from `searchsorted(..., right=True)` in [`earth2grid/_regrid.py`](https://github.com/NVlabs/earth2grid/blob/main/earth2grid/_regrid.py), which pushes an exact match at the upper endpoint past the last valid cell.

## Minimum reproducer

```python
import numpy as np
import torch
import earth2grid

source_grid = earth2grid.latlon.LatLonGrid(
    lat=[-30.0, 30.0],
    lon=[0.5, 90.5, 180.5, 270.5],
)

regridder = source_grid.get_bilinear_regridder_to(
    np.array([30.0], dtype=np.float32),
    np.array([1.0], dtype=np.float32),
).float()

x = torch.ones((1, 2, 4), dtype=torch.float32)
y = regridder(x)
print(y)  # tensor([[nan]])
```

## Observed behavior

```python
tensor([[nan]])
```

A nearby interior point is finite:

```python
regridder = source_grid.get_bilinear_regridder_to(
    np.array([29.999], dtype=np.float32),
    np.array([1.0], dtype=np.float32),
)
print(regridder(torch.ones((1, 2, 4), dtype=torch.float32)))
# tensor([[1.]])   # or very close
```

The lower boundary is also finite:

```python
regridder = source_grid.get_bilinear_regridder_to(
    np.array([-30.0], dtype=np.float32),
    np.array([1.0], dtype=np.float32),
)
print(regridder(torch.ones((1, 2, 4), dtype=torch.float32)))
# tensor([[1.]])
```

## Why this looks incorrect

For `y_coords = [-30, 30]` and `y_query = 30`, the current logic does:

```python
y_l_idx = torch.searchsorted(y_coords, y_query, right=True) - 1
y_u_idx = y_l_idx + 1
```

which gives:

- `y_l_idx = 1`
- `y_u_idx = 2`

Then the validity mask rejects the point because there is no row `2`, so the output remains `NaN`.

This makes the interpolation domain effectively half-open on the upper edge, even for exact matches to source coordinates.

## Expected behavior

One of these should happen:

- Exact endpoint queries should be accepted and mapped to the last valid cell, with weights giving the endpoint value.
- Or, if the intended contract is to reject the upper boundary, that should be documented explicitly, since it is surprising and asymmetric with the lower boundary.

## Environment

- `earth2grid` as used through `LatLonGrid.get_bilinear_regridder_to`
- PyTorch 2.x
- Python 3.12
